### PR TITLE
Docusaurus build framework + ingestion doc refresh.

### PR DESCRIPTION
**Pull request by gianm**
### TL;DR

A refresh of the documentation done in collaboration with @vogievetsky.

Check out a render at: https://staging-druid.imply.io/docs/design/

And compare to the current doc pages: https://druid.apache.org/docs/latest/design/

### Description

This refresh has two main goals.

First, setting up **Docusaurus**:

1. Use [Docusaurus](https://docusaurus.io/), a really nice documentation generator that creates much more beautiful pages than we have now. They are also more functional; each one has a left-hand collapsible table of contents showing the outline of the overall docs, and a right-hand table of contents showing the outline of that particular doc page. Compare to the old doc pages, where there's only one non-collapsible ToC, and it's grown so long that it's quite difficult to follow.
2. Make various titles and headers more consistent. The inconsistencies have been around for a while, but became much easier to notice with the new Docusaurus ToCs.

Second, an **ingestion doc refresh**:

1. Includes a new `ingestion/index.md` doc that introduces all the key ingestion spec concepts, and describes the most popular ingestion methods. It is meant to be an introduction to the world of Druid ingestion.
2. Consolidate lots of disparate ingestion docs, which have grown organically over time and have become difficult to follow, into a simpler set of fewer, larger, more cross-referenced pages. Check out the new pages at: https://staging-druid.imply.io/docs/ingestion/index.html. Some prime examples of this consolidation are the new pages `ingestion/data-management.md` and `ingestion-tasks.md`, which represent multiple pages from the current set of docs.
3. Have a bit more 'opinionatedness' in the ingestion pages, pushing new people towards Kafka, Kinesis, Hadoop, and native batch ingestion. They discuss Tranquility but don't present it as something highly recommended (in keeping with its status as something that doesn't get a lot of love these days). And the few remaining references to Realtime Nodes have been redirected to a page that says they're gone now.

### Other notes

I think we need to restore some of the `_bin` scripts that are still useful (but which ones?).

### CODE REVIEW

File: docs/dependencies/metadata-storage.md [hunk 1]
```diff
@@ -113,7 +110,7 @@ Note that the format of this blob can and will change from time-to-time.
 ### Rule Table
```
Comment (ccaominh) [inline docs/dependencies/metadata-storage.md]: The title case style here is not consistent with the changes you made above

File: docs/ingestion/index.md [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+| **Supervisor type** | `kafka` | `kinesis` | N/A |
+| **How it works** | Druid reads directly from Apache Kafka. | Druid reads directly from Amazon Kinesis. | Tranquility, a library that ships separately from Druid, is used to push data into Druid. |
+| **Can ingest late data?** | Yes | Yes | No (late data is dropped based on the `windowPeriod` config) |
+| **Exactly-once guarantees?** | Yes | Yes | No |
+
+### Batch
+
+When doing batch loads from files, you should use one-time [tasks](tasks.md), and you have three options: `index`
+(native batch; single-task), `index_parallel` (native batch; parallel), or `index_hadoop` (Hadoop-based). The following
+table compares and contrasts the three batch ingestion options.
```
Comment (ccaominh) [inline docs/ingestion/index.md]: This last sentence is slightly out of place. Perhaps merge it with the sentence on line 81?

File: docs/ingestion/index.md [hunk 3]
```diff
@@ -0,0 +1,769 @@
...
+By default, rollup is enabled.
+
+### Enabling or disabling rollup
+
+Rollup is controlled by the `rollup` setting in the [`granularitySpec`](#granularityspec). By default, it is `true`
+(enabled). Set this to `false` if you want Druid to store each record as-is, without any rollup summarization.
+
+### Example of rollup
+
+For an example of how to configure rollup, and of what how the feature will modify your data, check out the
```
Comment (ccaominh) [inline docs/ingestion/index.md]: Typo: of what how -> of how

File: docs/ingestion/index.md [hunk 4]
```diff
@@ -0,0 +1,769 @@
...
+that datasource leads to much faster query times. This can often be done with just a small increase in storage
+footprint, since abbreviated datasources tend to be substantially smaller.
+- If you are using a [best-effort rollup](#best-effort-rollup) ingestion configuration that does not guarantee perfect
+rollup, you can potentially improve your rollup ratio by switching to a guaranteed perfect rollup option, or by
+[reindexing](data-management.md#compaction-and-reindexing) your data in the background after initial ingestion.
+
+### Best-effort rollup
+
+Some Druid ingestion methods guarantee _perfect rollup_, meaning that input data are perfectly aggregated at ingestion
+time. Others offer _best-effort rollup_, meaming that input data might not be perfectly aggregated and thus there could
```
Comment (ccaominh) [inline docs/ingestion/index.md]: Typo: meaming -> meaning

File: docs/ingestion/index.md [hunk 5]
```diff
@@ -0,0 +1,769 @@
...
+be multiple segments holding rows with the same timestamp and dimension values.
+
+In general, ingestion methods that offer best-effort rollup do this because they are either parallelizing ingestion
+without a shuffling step (which would be required for perfect rollup), or because they are finalizing and publishing
+segments before all data for a time chunk has been received, which we call _incremental publishing_. In both of these
+cases, records may end up in different segments that are received by different, non-shuffling tasks cannot be rolled
```
Comment (ccaominh) [inline docs/ingestion/index.md]: This sentence needs to be reworded. Perhaps something like: non-shuffling tasks cannot be -> non-shuffling tasks and cannot be

File: docs/ingestion/index.md [hunk 6]
```diff
@@ -0,0 +1,769 @@
...
+This partitioning happens for all ingestion methods, and is based on the `segmentGranularity` parameter of your
+ingestion spec's `dataSchema`.
+
+The segments within a particular time chunk may also be partitioned further, using options that vary based on the
+ingestion method you have chosen. In general, doing this secondary partitioning using a particular dimension will
+improve locality, meaning that rows with the same value for that dimension are stored together and can be accessed
+quickly.
+
+You will usually get the best performance and smallest overall footprint by partitioning your data on some "natural"
+dimension that you often filter by, if one exists. This will often improve compression — users have reported threefold
```
Comment (ccaominh) [inline docs/ingestion/index.md]: Are the dashes rendered as em-dashes?

**Comment by vogievetsky** [comment]
Hi all, as @gianm mentioned this work is a collaboration. I am going to be pushing to this branch ( https://github.com/implydata/druid/tree/ingest-doc ) to address feedback and fix up some remaining things about the build script.

**Comment by fjy** [comment]
+2

**Comment by clintropolis** [comment]
>I think we need to restore some of the _bin scripts that are still useful (but which ones?).

#8306 moves all the scripts related to performing an apache release into `distribution/bin`, which makes more sense anyway since they are not doc related scripts.

This is all that remains in `docs/_bin`:
Screen Shot 2019-08-16 at 9 57 50 PM

which if none are necessary to the new docs can all safely be deleted I believe.

File: docs/ingestion/index.md (hunk 2 above)
Comment (gianm) [reply docs/ingestion/index.md]: I deleted it and replaced the L81 sentence with:

> This table compares the three available options:

File: docs/dependencies/metadata-storage.md (hunk 1 above)
Comment (gianm) [reply docs/dependencies/metadata-storage.md]: Fixed, thanks.

File: docs/ingestion/index.md (hunk 3 above)
Comment (gianm) [reply docs/ingestion/index.md]: Fixed, thanks.

File: docs/ingestion/index.md (hunk 4 above)
Comment (gianm) [reply docs/ingestion/index.md]: Fixed, thanks.

File: docs/ingestion/index.md (hunk 6 above)
Comment (gianm) [reply docs/ingestion/index.md]: Yes, this is rendered as an em dash.

File: docs/ingestion/index.md (hunk 5 above)
Comment (gianm) [reply docs/ingestion/index.md]: I replaced this sentence:

> In both of these cases, records may end up in different segments that are received by different, non-shuffling tasks cannot be rolled up together.

With this one:

> In both of these cases, records that could theoretically be rolled up may end up in different segments.

**Comment by gianm** [comment]
Just re-pushed.

File: website/sidebars.json [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+      "querying/multi-value-dimensions",
+      "querying/lookups",
+      "querying/joins",
+      "querying/multitenancy",
+      "querying/caching",
+      "development/geo"
+    ],
+    "Configuration": [
+      "configuration/index",
+      "development/extensions",
```
Comment (yurmix) [inline website/sidebars.json]: Very nice outline! Also, linking straight to .md is great.

One thing though: I see that the list of extensions (extensions.md) has been moved under Configuration. I think it's better to mention in the release note this change and any other structural change, to let existing users know and adjust.

**Comment by yurmix** [comment]
Do you think you can find a way to control style/indentation for subcategories on the sidebar? I think the distinction between 2nd and 3rd levels (and so on) isn't clear enough. 
The only thing I found is subcategory collapse, which should be available on v2 (currently early alpha): https://github.com/facebook/Docusaurus/issues/1352

**Comment by vogievetsky** [comment]
@yurmix I looked into Docusaurus v2 and it looks like it is not quite ready for prime time use which is why we did not use it here. Totally down to switch to it once it comes out (or at least becomes more stable).

As for the indentation we can play with the CSS a bit to make it more obvious in the meantime - in a subsequent PR, trying to get this merged so we do not have to live in conflict hell.

File: website/sidebars.json (hunk 7 above)
Comment (gianm) [reply website/sidebars.json]: Ah yes, I did move it there, since I thought it made more sense there. (The Extensions page is written in a user-facing style, and so I don't think it belonged in the Development section.)

File: docs/design/processes.md [hunk 8]
```diff
@@ -88,18 +88,30 @@ Within a Data server, functionality is split between two processes, the Historic
...
 for reading from external data sources and publishing new Druid segments.
 
 #### Peon processes
 
-[**Peon**](../design/peons.html) processes are task execution engines spawned by MiddleManagers. Each Peon runs a separate JVM and is responsible for executing a single task. Peons always run on the same host as the MiddleManager that spawned them.
+[**Peon**](../design/peons.md) processes are task execution engines spawned by MiddleManagers. Each Peon runs a separate JVM and is responsible for executing a single task. Peons always run on the same host as the MiddleManager that spawned them.
+
+### Indexer process (optional)
+
+[**MiddleManager**](../design/indexer.md) processes are an alternative to MiddleManagers and Peons. Instead of
```
Comment (jon-wei) [inline docs/design/processes.md]: MiddleManager -> Indexer

File: docs/ingestion/index.md [hunk 9]
```diff
@@ -0,0 +1,768 @@
...
+
+In this query, `cnt` should refer to a "count" type metric specified at ingestion time. See
+[Counting the number of ingested events](schema-design.md#counting) on the "Schema design" page for more details about
+how counting works when rollup is enabled.
+
+Tips for maximizing rollup:
+
+- Generally, the fewer dimensions you have, and the lower the cardinality of your dimensions, the better rollup ratios
+you will achieve.
+- Use [sketches](#sketches) to avoid storing high cardinality dimensions, which harm rollup ratios.
```
Comment (jon-wei) [inline docs/ingestion/index.md]: The `#sketches` link doesn't resolve

File: docs/ingestion/index.md [hunk 10]
```diff
@@ -0,0 +1,768 @@
...
+Transforms do have some limitations. They can only refer to fields present in the actual input rows; in particular,
+they cannot refer to other transforms. And they cannot remove fields, only add them. However, they can shadow a field
+with another field containing all nulls, which will act similarly to removing the field.
+
+Transforms can refer to the [timestamp](#timestamp) of an input row by referring to `__time` as part of the expression.
+They can also _replace_ the timestamp if you set their "name" to `__time`. In both cases, `__time` should be treated as
+a millisecond timestamp (number of milliseconds since Jan 1, 1970 at midnight UTC). Transforms are applied _after_ the
+`timestampSpec`.
+
+Druid currently includes one kind of builtin transform, the expression transform. It has the following syntax:
```
Comment (jon-wei) [inline docs/ingestion/index.md]: builtin -> built-in

File: docs/ingestion/native-batch.md [hunk 11]
```diff
@@ -487,10 +493,13 @@ If you have some tasks of a higher priority than others, you may set their
 This may help the higher priority tasks to finish earlier than lower priority tasks
 by assigning more task slots to them.
 
-Local Index Task
-----------------
+## Simple task
```
Comment (jon-wei) [inline docs/ingestion/native-batch.md]: :+1: on the new name

File: docs/comparisons/druid-vs-elasticsearch.md [hunk 12]
```diff
@@ -22,19 +22,18 @@ title: "Apache Druid (incubating) vs Elasticsearch"
...
   -->
 
-# Apache Druid (incubating) vs Elasticsearch
 
 We are not experts on search systems, if anything is incorrect about our portrayal, please let us know on the mailing list or via some other means.
 
-Elasticsearch is a search systems based on Apache Lucene. It provides full text search for schema-free documents 
-and provides access to raw event level data. Elasticsearch is increasingly adding more support for analytics and aggregations. 
-[Some members of the community](https://groups.google.com/forum/#!msg/druid-development/nlpwTHNclj8/sOuWlKOzPpYJ) have pointed out  
+Elasticsearch is a search systems based on Apache Lucene. It provides full text search for schema-free documents
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/comparisons/druid-vs-elasticsearch.md]: Fixed, thanks.

File: docs/design/processes.md (hunk 8 above)
Comment (gianm) [reply docs/design/processes.md]: Fixed, thanks.

File: docs/ingestion/index.md (hunk 9 above)
Comment (gianm) [reply docs/ingestion/index.md]: Fixed, thanks.

File: docs/ingestion/index.md (hunk 10 above)
Comment (gianm) [reply docs/ingestion/index.md]: Fixed, thanks.

File: docs/design/index.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+3. **Massively parallel processing.** Druid can process a query in parallel across the entire cluster.
+4. **Realtime or batch ingestion.** Druid can ingest data either real-time (ingested data is immediately available for
+querying) or in batches.
+5. **Self-healing, self-balancing, easy to operate.** As an operator, to scale the cluster out or in, simply add or
+remove servers and the cluster will rebalance itself automatically, in the background, without any downtime. If any
+Druid servers fail, the system will automatically route around the damage until those servers can be replaced. Druid
+is designed to run 24/7 with no need for planned downtimes for any reason, including configuration changes and software
+updates.
+6. **Cloud-native, fault-tolerant architecture that won't lose data.** Once Druid has ingested your data, a copy is
+stored safely in [deep storage](#deep-storage) (typically cloud storage, HDFS, or a shared filesystem). Your data can be
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/design/index.md]: Fixed, thanks.

File: docs/ingestion/index.md [hunk 14]
```diff
@@ -0,0 +1,768 @@
...
+|---|-----|--------------|------------|
+| **Task type** | `index` | `index_parallel` | `index_hadoop` |
+| **Parallel?** | No. Each task is single-threaded. | Yes, if firehose is splittable and `maxNumConcurrentSubTasks` > 1 in tuningConfig. See [firehose documentation](native-batch.md#firehoses) for details. | Yes, always. |
+| **Can append or overwrite?** | Yes, both. | Yes, both. | Overwrite only. |
+| **External dependencies** | None. | None. | Hadoop cluster (Druid submits Map/Reduce jobs). |
+| **Input locations** | Any [firehose](native-batch.md#firehoses). | Any [firehose](native-batch.md#firehoses). | Any Hadoop FileSystem or Druid datasource. |
+| **File formats** | Text file formats (CSV, TSV, JSON). Support for binary formats is coming in a future release. | Text file formats (CSV, TSV, JSON). Support for binary formats is coming in a future release. | Any Hadoop InputFormat. |
+| **[Rollup modes](#rollup)** | Perfect if `forceGuaranteedRollup` = true in the [`tuningConfig`](native-batch.md#tuningconfig).| Only best-effort. Support for perfect rollup is coming in a future release. | Always perfect. |
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/ingestion/index.md]: I updated the "rollup modes" and "partitioning options" sections.

File: docs/ingestion/index.md [hunk 15]
```diff
@@ -0,0 +1,768 @@
...
+flexibility. As of current Druid versions, If you are doing initial ingestion through a less-flexible method (like
+Kafka) then you can use [reindexing techniques](data-management.html#compaction-and-reindexing) to repartition your data after it
+is initially ingested. This is a powerful technique: you can use it to ensure that any data older than a certain
+threshold is optimally partitioned, even as you continuously add new data from a stream.
+
+The following table shows how each ingestion method handles partitioning:
+
+|Method|How it works|
+|------|------------|
+|[Native batch](native-batch.html)|`index` (non-parallel) tasks partition input files based on the `partitionDimensions` and `forceGuaranteedRollup` tuning configs. `index_parallel` tasks do not currently support user-defined partitioning.|
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/ingestion/index.md]: I updated it to say:

```
Configured using [`partitionsSpec`](native-batch.html#partitionsspec) inside the `tuningConfig`.
```

File: docs/ingestion/data-management.md [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+### Replacing segments
+
+Druid uniquely
+identifies segments using the datasource, interval, version, and partition number. The partition number is only visible in the segment id if
+there are multiple segments created for some granularity of time. For example, if you have hourly segments, but you
+have more data in an hour than a single segment can hold, you can create multiple segments for the same hour. These segments will share
+the same datasource, interval, and version, but have linearly increasing partition numbers.
+
+```
+foo_2015-01-01/2015-01-02_v1_0
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/ingestion/data-management.md]: Maybe, I'd rather change that later though (if we do at all) since this section was just copied and relocated from an existing document.

File: docs/ingestion/index.md [hunk 17]
```diff
@@ -0,0 +1,768 @@
...
+
+Ingestion methods that guarantee perfect rollup do it with an additional preprocessing step to determine intervals
+and partitioning before the actual data ingestion stage. This preprocessing step scans the entire input dataset, which
+generally increases the time required for ingestion, but provides information necessary for perfect rollup.
+
+The following table shows how each method handles rollup:
+
+|Method|How it works|
+|------|------------|
+|[Native batch](native-batch.html)|`index_parallel` type is best-effort. `index` type may be either perfect or best-effort, based on configuration.|
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/ingestion/index.md]: Fixed, thanks.

File: docs/ingestion/native-batch.md [hunk 18]
```diff
@@ -73,16 +78,19 @@ if one of them fails.
 
 You may want to consider the below things:
 
+- This task does not support [perfect rollup](index.md#best-effort-rollup) because it does not shuffle
```
[the remark this answers is not in the document]
Comment (gianm) [reply docs/ingestion/native-batch.md]: Fixed, thanks.

**Comment by gianm** [comment]
There are a bunch of broken anchor-links I'll look at fixing later today. (Found by running `npm run anchor-lint`)

**Comment by gianm** [comment]
I've pushed an update reflecting the comments above, with all broken links and anchors fixed, and re-uploaded a render to https://staging-druid.imply.io/docs/design/index.html. I've also restored some subheaders that were accidentally deleted.

**Comment by gianm** [comment]
> I've pushed an update reflecting the comments above, with all broken links and anchors fixed, and re-uploaded a render to https://staging-druid.imply.io/docs/design/index.html. I've also restored some subheaders that were accidentally deleted.

I spoke too soon; the broken link checker built into docusaurus wasn't catching all the broken links (it only checked .md links, not .html links). Will need to fix this and re-push.

File: .travis.yml [hunk 19]
```diff
@@ -164,6 +164,10 @@ matrix:
       script:
         - $MVN test -pl 'web-console'
 
+    - name: "docs"
+      install: cd website && npm install
+      script: cd website && npm run lint
```
Comment (ccaominh) [inline .travis.yml]: From the travis log, it looks like the `cd website` done in the `install` step stays in effect when the `script step`.  Some alternatives are to use pushd/popd or a subshell.

**Comment by gianm** [comment]
Pushed again to resolve some conflicts, fixed more broken links and anchors found by the new script `npm run lint`, and uploaded another render (including redirects!) here: https://staging-druid.imply.io/docs/

File: docs/ingestion/index.md [hunk 20]
```diff
@@ -0,0 +1,769 @@
...
+>
+> For more details on splitting data up into separate datasources, and potential operational considerations, refer
+> to the [Multitenancy considerations](../querying/multitenancy.md) page.
+
+<a name="spec"></a>
+
+## Ingestion specs
+
+No matter what ingestion method you use, data is loaded into Druid using either one-time [tasks](tasks.html) or
+ongoing [supervisors](supervisors.html). In any case, part of the task or supervisor definition is an
```
Comment (jihoonson) [inline docs/ingestion/index.md]: Broken link.

File: docs/ingestion/data-formats.md [hunk 21]
```diff
@@ -56,18 +54,18 @@ _CSV_
...
 
 Note that the CSV and TSV data do not contain column heads. This becomes important when you specify the data for ingesting.
 
 ## Custom Formats
 
-Druid supports custom data formats and can use the `Regex` parser or the `JavaScript` parsers to parse these formats. Please note that using any of these parsers for 
+Druid supports custom data formats and can use the `Regex` parser or the `JavaScript` parsers to parse these formats. Please note that using any of these parsers for
 parsing data will not be as efficient as writing a native Java parser or using an external stream processor. We welcome contributions of new Parsers.
 
 ## Configuration
```
Comment (jihoonson) [inline docs/ingestion/data-formats.md]: This is not introduced in this pr, but would you please fix the broken link [below](https://github.com/apache/incubator-druid/pull/8311/files#diff-57f00ca443fbdb6178e33d1029f00817R89) as well? `Druid can automatically flatten it for you` is broken.

**Comment by vogievetsky** [comment]
OMG! it looks good. It even passed my super strict broken link linter! Someone merge before there are more doc conflicts.

File: docs/ingestion/index.md (hunk 20 above)
Comment (gianm) [reply docs/ingestion/index.md]: Fixed, thanks.

File: docs/ingestion/data-formats.md (hunk 21 above)
Comment (gianm) [reply docs/ingestion/data-formats.md]: Thanks, fixed this too (and now it's being detected by the broken link checker).

**Comment by leventov** [comment]
Did you use some regular expression to replace `.html` -> `.md` links? Is that on purpose that links with id (like `../configuration/index.html#dynamic-configuration`) remain with `.html`?

**Comment by gianm** [comment]
> Did you use some regular expression to replace `.html` -> `.md` links? Is that on purpose that links with id (like `../configuration/index.html#dynamic-configuration`) remain with `.html`?

The `.md` links are a little nicer since they will make links work when the source is viewed in GitHub, and will generate to the exact same link on the live site. I think it'd be fine to replace the ones with anchor ids too. I think I replaced a few by hand but didn't do all of them.
